### DISCUSSION TITLE
How can I trigger a function on the frame (only the frame) the player hits an obstacle?

### DISCUSSION THREAD

**Discussion by TechStudent10** [geode-sdk/sdk#107] [Help] [answered]
To create a NoClip deaths mod, I've been using the `PlayLayer::destroyPlayer` function to override it's functionality. However, as I've discovered, `destroyPlayer` is called every frame the player is colliding with an obstacle.

How may I overcome this?

**Comment by HJfod** [comment, marked as answer]
I don't know the signature of `destroyPlayer` and I don't know exactly how to do it, but the way I'd approach figuring it out is probably by first checking to see if there exists a function for that (probably doesn't as in normal gameplay there's no reason to store whether the player has collided or not as the player will die if any collision is detected).

If a function specifically for it doesn't exist, then I'd first figure out what exactly the functionality you need is.

If you only need to check if the player would've died anywhere in the level, I'd just store a global boolean flag for that, set it in `destroyPlayer` and then reset at level start.

If your goal is to count every time the player would've died, I can think of two ways to approach that.

The first idea that comes to mind is to check which object the player is currently colliding with in `destroyPlayer` and store those in a `std::unordered_set`, then check if an object is already in that set to see if the player would die to something else.

However, if the goal is counting "deaths", you should realize that that's not exactly a well-defined thing if the player doesn't actually die. Does counting deaths mean every frame that the player would've died counts as a death? If so, that's the default behaviour of `destroyPlayer`. However, judging by your post, that's not ideal for what you want.

If you want a death count that's easy to digest for the user, I'd say the best way to do it is to count deaths by every frame the player would die, but skip consecutive frames. As in, when `destroyPlayer` is first called, increment the death count and set a global flag for "player is currently dying", and then check every frame if the player is still dying, and if yes, don't increment the death count and if no, set the flag back to the default state so the next time the player starts dying is calculated. This way, if the player fails to jump over triple spikes or something equivalent, that only counts as one death.

To implement that, all you need is some way to check if the player is no longer dying. One way to do this would be to find which function is called every frame to check for death (probably `PlayerObject::update`) and then use flags to detect if `destroyPlayer` was called during that frame. Hooking a function that is called every frame can incur a noticable performance drop though.

**Reply by TechStudent10** [reply]
Ah thank you! I don't really that much about performance because the player probably won't have it enabled at all times.

**Reply by TechStudent10** [reply]
I have code that looks like this:

```cpp
class $modify(PlayerObject) {
	void update() {
		if (collidedWithObject()) {}
	}
};
```
(empty for now)

VSCode keeps bugging me with this error and I have no clue where to start (I'm not an avid C++ user. I use it when I have to):

![image](https://user-images.githubusercontent.com/76978184/210288821-ca7acd36-cccb-4c06-862f-1ecf1280ee72.png)

I got the `collidedWithObject` function from https://github.com/geode-sdk/geode/blob/main/bindings/GeometryDash.bro#L4479/

**Reply by altalk23** [reply]
First is that update function takes a float parameter dt (the amount of time passed), you need to add that
Second is that collided with object has 2 overloads, `bool collidedWithObject(float, GameObject*)` and `bool collidedWithObject(float, GameObject*, cocos2d::CCRect)`
honestly idk what the float and ccrect params are but gameobject* param is the object you are checking collision with, you will proabbly need to reverse engineer to find what the other params are

**Reply by altalk23** [reply]
and on top of that im not sure if this is the function you want either, you should probably not do checks on every object every frame (that would be bad for your fps)

**Reply by TechStudent10** [reply]
> and on top of that im not sure if this is the function you want either, you should probably not do checks on every object every frame (that would be bad for your fps)

Ah okay. I'll continue doing my research.

**Reply by TechStudent10** [reply]
After trying a new approach, my mod won't compile because of this:

> Sorry if this is such a nooby question

```cpp
class $modify(MyPlayLayer, PlayLayer) {
	int m_deaths;
	bool m_incrementDeaths;
	bool m_isColiding;

	MyPlayLayer() : m_deaths(0) {}
	MyPlayLayer() : m_incrementDeaths(false) {}
	MyPlayLayer() : m_isColiding(false) {}
// ...
```

This line, `MyPlayLayer() : m_deaths(0) {}` works fine, but the other two don't. All I'm doing is following documentation.

**Reply by HJfod** [reply]
What you have done there is define three different constructors that take identical parameters but have different variables initialized. Initialize all the members in one constructor:
```cpp
MyPlayLayer() : m_deaths(0), m_incrementDeaths(false), m_isColiding(false) {}
```
Btw, `isColiding` is misspelt :-)

**Reply by TechStudent10** [reply]
> What you have done there is define three different constructors that take identical parameters but have different variables initialized. Initialize all the members in one constructor:
> ```cpp
> MyPlayLayer() : m_deaths(0), m_incrementDeaths(false), m_isColiding(false) {}
> ```
> Btw, `isColiding` is misspelt :-)

Ah thank you! (I totally noticed the misspelling)
